source/java/src/main/java/com/uber/neuropod/NeuropodTensor.java

```diff
@@ -0,0 +1,131 @@
+package com.uber.neuropod;
+
+import java.util.List;
+
+/**
+ * The interface Neuropod tensor, which stores the data of a tensor. NeuropodTensors are backed
+ * by a buffer, and the buffer may be allocated by the C++ side(inference result) or the Java side.
+ */
+interface NeuropodTensor {
+     /**
+      * Get the dims array which represnents the shape of a tensor.
+      *
+      * @return the shape array
+      */
+     long[] getDims();
+
+     /**
+      * Gets number of elements of a tensor.
+      *
+      * @return the number of elements
+      */
+     long getNumberOfElements();
+
+     /**
+      * Gets type of a tensor
+      *
+      * @return the tensor type
+      */
+     TensorType getTensorType();
+
+     /**
+      * Flatten the tensor data and convert it to a long array.
+      *
+      * Can only be used when the tensor is INT64_TENSOR. Will trigger
+      * a copy if the backed buffer is not a LongBuffer backed by an array.
+      *
+      * @return the long [ ]
+      */
+     long[] toLongArray();
+
+     /**
+      * Flatten the tensor data and convert it to a int array.
+      *
+      * Can only be used when the tensor is INT32_TENSOR. Will trigger
+      * a copy if the backed buffer is not a IntBuffer backed by an array.
+      *
+      * @return the int [ ]
+      */
+     int[] toIntArray();
+
+     /**
+      * Flatten the tensor data and convert it to a float array.
+      *
+      * Can only be used when the tensor is FLOAT_TENSOR. Will trigger
+      * a copy if the backed buffer is not a FloatBuffer backed by an array.
+      *
+      * @return the float [ ]
+      */
+     float[] toFloatArray();
+
+     /**
+      * Flatten the tensor data and convert it to a double array.
+      *
+      * Can only be used when the tensor is DOUBLE_TENSOR. Will trigger
+      * a copy if the backed buffer is not a DoubleBuffer backed by an array.
+      *
+      * @return the double [ ]
+      */
+     double[] toDoubleArray();
```

Comment:
Doesn't match the RFC (and as we talked about this won't work without copies for direct ByteBuffers). We should replace these with methods that return `IntBuffer`, `LongBuffer`, `FloatBuffer`, and `DoubleBuffer`.

Ideally we wouldn't have all of these methods on `NeuropodTensor` because most of them won't be valid on any given tensor. For now, it should be fine, but we should think about other approaches in the future. In C++, we have a templated subclass of `NeuropodTensor` called `TypedNeuropodTensor<T>` where T can be any of the supported tensor types.

Not sure how easily we can do something similar with generics in Java since we can't do things like specializations. Maybe a more straightforward approach would be to create a subclass for each type (e.g. `FloatNeuropodTensor`, `StringNeuropodTensor`, etc.).

Reply:
The problem is generic types for primitive types is not supported in Java, so having different functions for different types seems to be a common solution(although for certain tensor only one function may apply). See the tensorflow Java API for more details:[](https://www.tensorflow.org/api_docs/java/reference/org/tensorflow/Tensor#writeTo(java.nio.LongBuffer))